tgui/packages/tgui/components/NumberInput.tsx

```diff
@@ -0,0 +1,351 @@
+import { KEY } from 'common/keys';
+import { clamp } from 'common/math';
+import { BooleanLike, classes } from 'common/react';
+import {
+  Component,
+  createRef,
+  FocusEventHandler,
+  KeyboardEventHandler,
+  MouseEventHandler,
+  RefObject,
+} from 'react';
+
+import { AnimatedNumber } from './AnimatedNumber';
+import { Box } from './Box';
+
+const DEFAULT_UPDATE_RATE = 400;
+
+type NumberInputProps = {
+  value: number | string;
+  minValue: number;
+  maxValue: number;
+  step?: number;
+  stepPixelSize?: number;
+  suppressFlicker?: number;
+  disabled?: BooleanLike;
+
+  className?: string;
+  fluid?: BooleanLike;
+  animated?: BooleanLike;
+  unit?: string;
+  height?: string;
+  width?: string;
+  lineHeight?: string;
+  fontSize?: string;
+  updateRate?: number;
+  format?: (value: number) => string;
+  onChange?: (e: any, value: number) => void;
+  onDrag?: (e: any, value: number) => void;
+};
```

Comment:
there's a helper for this that can make it neater, and we should use it to avoid chaining so many ?s

```tsx
type Props = {
... 
} & Partial<{ // optionals
step: number;
...
};
```

Reply:
I have a problem. after using partial i ran into this
![Screenshot (397)](https://github.com/tgstation/tgstation/assets/110812394/75c4c106-cb37-45bb-aeea-ed8a5838a43f)
How do i declare prop types?

Reply:
Never mind. Fixed it by using `Require<{}> & Partial<{}>`

Reply:
This does look like a bug! Does it still give a compiler error if you type it as
defaultProps: Props?

Reply:
No luck :( .